module/zfs/dmu_send.c

```diff
@@ -2664,20 +2685,53 @@ dmu_send(const char *tosnap, const char *fromsnap, boolean_t embedok,
 	dspp.resumeobj = resumeobj;
 	dspp.resumeoff = resumeoff;
 	dspp.rawok = rawok;
+	dspp.savedok = savedok;
 
 	if (fromsnap != NULL && strpbrk(fromsnap, "@#") == NULL)
 		return (SET_ERROR(EINVAL));
 
 	err = dsl_pool_hold(tosnap, FTAG, &dspp.dp);
 	if (err != 0)
 		return (err);
+
 	if (strchr(tosnap, '@') == NULL && spa_writeable(dspp.dp->dp_spa)) {
 		/*
 		 * We are sending a filesystem or volume.  Ensure
 		 * that it doesn't change by owning the dataset.
 		 */
-		err = dsl_dataset_own(dspp.dp, tosnap, dsflags, FTAG,
-		    &dspp.to_ds);
+		if (savedok) {
+			char name[ZFS_MAX_DATASET_NAME_LEN + 6];
+
+			(void) snprintf(name, sizeof (name),
+			    "%s/%%recv", tosnap);
```

Comment:
Should probably use `recv_clone_name` constant instead of `%%recv` (and compute the size of `name` dependent on `recv_clone_name`'s length)

https://github.com/zfsonlinux/zfs/blob/caf9dd209fdcfccabc2f32b3f23c5386ccfb896c/module/zfs/dmu_recv.c#L66

Reply:
I think we allocate ZFS_MAX_DATASET_NAME_LEN  on the stack in several other places, but it looks like you could easily use `kmem_asprintf()` to allocate it on the heap here.